Harmonic hosting capacity

hakanergun opened the pull request:
Versions:
v0.1.0 - init, hosting capacity with fixed angle of the harmonic injection.
v0.2.0 - allow for variable angle of the harmonic injection, to enable (partial) recombination of the harmonic pollution.

timmyfaraday committed:
first working version hhc

hakanergun committed:
quadratic objective

timmyfaraday committed:
update input for ihd and thd limits

The IHD and THD limits can now be provided through setting the applicable standard at a given bus. Currently, implemented standards are IEC61000-2-4:2002, Cl. 2 and IEC61000-3-6:2008.

timmyfaraday committed:
Merge branch 'test_hhc' of https://github.com/timmyfaraday/HarmonicPowerModels.jl into test_hhc

timmyfaraday committed:
small changes

timmyfaraday committed:
small changes

hakanergun committed:
Added harmonic current injection angle range

- variable hci (harmonic current injection) for each load
- adding conic contraint hci^2 <= crd^ + cid^2

hakanergun committed:
fix circular functions

change cosine to tangent function

timmyfaraday committed:
review variable angle

added features:
- reset error where fundamental load power was set to zero
- changed tan to sin and cos, with min/max operator for angle logic
- cleaned up nomenclature

timmyfaraday commented:
### Version 0.3.0 - Harmonic Fridays 24/11
- [x] Make slides for the monthly meeting, show v0.2.0, i.e., deterministic problem is ready
- [x] Check failing tests of the package
- [x] Have a discussion on slack bus behavior for harmonic currents, i.e., behavior of (29) and (30) in v0.2.0
- [x] Build v0.3.0: SOC problem formulation ready for robust optimization, including assumptions:
1. Planning problem for fundamental component is out of scope
2. Fundamental voltage becomes a parameter, in future versions, a stochastic parameter
-> Concretely, this drops the lower bound in (7), and entirely drops (31) and (32) in v0.2.0
- [ ] Relate the harmonic current angle cone to the general summation law in IEC61000-3-6, see: Extension on IEC 61000-3-6’s General Summation Law to Estimate Harmonic Current at PCC Based on Probability Distribution by Trelles-Molina et al.

hakanergun committed:
script for analysing angle ranges

timmyfaraday committed:
fix error idh

timmyfaraday committed:
addition of tests, not yet running through

timmyfaraday committed:
updated test -- runs through locally

hakanergun committed:
adding constarint for ref angle based on nodal voltage angle.

hakanergun committed:
adding vector group shift

hakanergun committed:
Merge branch 'test_hhc' of https://github.com/timmyfaraday/HarmonicPowerModels.jl into test_hhc

timmyfaraday committed:
Update hhc.jl

timmyfaraday committed:
Merge branch 'test_hhc' of https://github.com/timmyfaraday/HarmonicPowerModels.jl into test_hhc

hakanergun committed:
add phasor plots using ElectricalEngineering.jl

hakanergun committed:
Merge branch 'test_hhc' of https://github.com/timmyfaraday/HarmonicPowerModels.jl into test_hhc

hakanergun committed:
bug fix plot

timmyfaraday commented:
### Version 0.3.0 -- Harmonic Fridays 8/12

- [x] Add Type to dispatch to specific formulation (@timmyfaraday)
- [x] Add SOC formulation v0.3.0 (@timmyfaraday)
- [ ] Add automatic vector group shift wrt reference bus (@hakanergun)
- [ ] Add c_rating of the load to the input file, not a-postiori (@hakanergun)
- [x] Add slack bus as discussed 1/12 (@timmyfaraday), no direct way to set the reference angle harmonic voltage distortion
- [x] Remove ElectricalEngineering from Pkg, otherwise this has to built and downloaded everytime the tests are run (@hakanergun)

timmyfaraday committed:
add short circuit impedance

timmyfaraday committed:
attempt at qc and soc

- closes pt 1 of v0.3.0
- closes pt 2 of v0.3.0

hakanergun committed:
Conic SOC implemented

- Fixing conic constraints for current and thd
- Adding rotated cone for harmonic current limits for the reference angle / range constraints.
- Deactivating NL expressions for post-processing

timmyfaraday committed:
if clause to avoid NLexpressions for SOC form

timmyfaraday committed:
update soc constraints

timmyfaraday commented:
### Version 0.3.0 -- Harmonic Fridays 15/12 

- [ ] Have a discussion on how to handle RMS limit for current through a branch
- [ ] Evaluate flows

timmyfaraday committed:
two_branch_example_hhc

timmyfaraday committed:
Update iv.jl

timmyfaraday commented:
# HARMONIC - January
Goal should be to build, document and validate the two models for the deterministic harmonic hosting capacity problem:
- NLP (`dHHC_NLP`): exact formulation, and
- SOC (`dHHC_SOC`): approximation of the load current magnitude.

The following grid elements are currently considered: 1) slack bus, 2) branches (pi-model), 3) transformers (t-model), 4) passive shunt elements, and 5) general harmonic load where for h=1: constant power and for h>1: constant current.

## TO DOs 
### Build models
- [x] Reference bus with pre-set harmonic distortion (@timmyfaraday by 19/01)
- [x] Add relevant bounds to all variables (@hakanergun by 19/01)
- [ ] Add cut to strengthen the SOC approximation (on 02/02)
- [x] Add current limits to the transformers (@timmyfaraday)

### Validate models
- [ ] Validate the flows through the industrial network (@timmyfaraday + @hakanergun on 19/01)
- [ ] Validate the necessity of the load current angle cone (@timmyfaraday + @hakanergun on 19/01)
An alternative could be to apply the general summation law directly.

### Document models
- [x] Build two overleaf documents detailing the models (@timmyfaraday)
- [x] Build slide set detailing the features of the models, including considered grid elements (@timmyfaraday by 26/01)

timmyfaraday committed:
reference node constraint with preset harmonic voltage distortion

timmyfaraday committed:
addition of xfmr current rms limit

timmyfaraday committed:
spring cleaning dHHC_NLP

timmyfaraday committed:
small changes

timmyfaraday committed:
runs through to infeasibility

timmyfaraday committed:
spring cleaning SOC PM type

timmyfaraday committed:
fixed infeasibility

hakanergun committed:
Fix magentization constraint SOC

hakanergun committed:
add missing constraints SOC

hakanergun committed:
Update hhc.jl

Remove generator bounds for SOC

hakanergun committed:
Update industrial_network_hhc.jl

hakanergun committed:
Add current bounds for branch and transformer currents

- Also use own variable definitions for SOC instead of PMs for currents and voltages.

timmyfaraday committed:
update with running tests

timmyfaraday committed:
update of the pkgs and documentation

timmyfaraday committed:
Update README.md

timmyfaraday committed:
Update README.md

timmyfaraday committed:
Update Project.toml

timmyfaraday committed:
Merge branch 'test_hhc' of https://github.com/timmyfaraday/HarmonicPowerModels.jl into test_hhc

timmyfaraday committed:
Update README.md

timmyfaraday committed:
additions

- documentation files
- dHHC_SOC: rewrite to include fundamental for the data to enable a preceeding fundamental power flow, but exclude it in the harmonic hosting capacity problem

hakanergun committed:
adding bounds

hakanergun committed:
Merge branch 'test_hhc' of https://github.com/timmyfaraday/HarmonicPowerModels.jl into test_hhc

hakanergun committed:
set bounded to true

hakanergun committed:
fix transformer voltage bounds

hakanergun committed:
fix current and voltage bounds for transformers

hakanergun committed:
Add current rating for generators

timmyfaraday committed:
hard code fundamental

hakanergun committed:
solve fundamental PF and write its results for SOC

timmyfaraday committed:
fundamental voltages and currents added for the SOC form

timmyfaraday committed:
Merge branch 'test_hhc' of https://github.com/timmyfaraday/HarmonicPowerModels.jl into test_hhc

hakanergun committed:
working SOC with fundamental components

timmyfaraday commented:
# HARMONIC - February
Two models for the deterministic harmonic hosting capacity problem:
- NLP (`dHHC_NLP`): exact formulation, and
- SOC (`dHHC_SOC`): approximation of the load current magnitude.

The following grid elements are currently considered: 1) slack bus, 2) branches (pi-model), 3) transformers (t-model), 4) passive shunt elements, and 5) general harmonic load where for h=1: constant power and for h>1: constant current.

## TO DOs to get to v0.2.0
### Addition of features
- [x] Add fairness principle to the objective (@timmyfaraday on 15/02)
- [x] Addition of four fairness principles in the paper (@timmyfaraday on 29/02)

### Validate models
- [x] Validate the flows through the industrial network (@timmyfaraday + @hakanergun on 16/02)

### Document models (after pushing v0.2.0 to main)
- [ ] Build documentation on Github pages (@timmyfaraday)

### Code review
- [x] Clean implementation of active filter (@timmyfaraday on 15/02)
- [x] General spring cleaning (@timmyfaraday + @hakanergun on 15-16/02)
- [x] Fix issues with HOPF implementation (@timmyfaraday + @frederikgeth on 20/02)

## TO DOs to get to v0.2.1
### Recombination of Harmonic Currents 
Investigate methods to enable recombination of harmonic currents from different harmonic loads: current angle cone vs general summation law
- [ ] Implement method for recombination of harmonic currents (@timmyfaraday + @hakanergun)
- [ ] If current angle cone, add cut to strengthen the SOC approximation (@timmyfaraday + @hakanergun)

## TO DOs for first journal paper on deterministic harmonic hosting capacity

- [ ] Introduction - Background and Motivation (@hakanergun)
- [x] Mathematical Framework - Harmonic Power Flow (@timmyfaraday)
- [x] Mathematical Framework - Harmonic Hosting Capacity (@timmyfaraday)

timmyfaraday committed:
spring cleaning solve_hhc

hakanergun committed:
unit tests for SOC using Hypatia

hakanergun committed:
Merge branch 'test_hhc' of https://github.com/timmyfaraday/HarmonicPowerModels.jl into test_hhc

timmyfaraday committed:
addition of filter category

1) start of spring cleaning
2) addition of filter category
- addition of add_ref_filter
- addition of :crf and :cif variables
- reformulation of active filter constraint

timmyfaraday committed:
spring cleaning - new headers

timmyfaraday committed:
spring cleaning - rm general import of functions

timmyfaraday committed:
update CI to julia v1.9

timmyfaraday committed:
spring cleaning -  hhc_soc init hpf

timmyfaraday committed:
fairness principle - equality

@hakanergun Getting weird behavior where all load current magnitudes for all loads and all harmonic equal each other, whereas only load current magnitudes for all loads and a given harmonic should equal each other.
Something to be reviewed during HF 16/02

timmyfaraday committed:
add fairness principle - updated tests

timmyfaraday committed:
spring cleaning - fairness principle

issues with fairness principle seems to be gone

hakanergun committed:
Validation of branch and transformer currents

timmyfaraday committed:
added test for hhc - ongoing work

timmyfaraday committed:
added test, has error for zero seq voltage angle shift when transfo has two YZD windings

timmyfaraday committed:
spring cleaning towards v0.2

- all references to transformer, t, etc. are replaced by xfmr, x, etc.
- updated header

timmyfaraday committed:
industrial_network_hopf fixed

hakanergun committed:
Adding bounds for NLP HHC

timmyfaraday committed:
small changes

timmyfaraday committed:
Merge branch 'test_hhc' of https://github.com/timmyfaraday/HarmonicPowerModels.jl into test_hhc

hakanergun committed:
add sqrt(3) in branch rating

hakanergun committed:
fix .toml

hakanergun committed:
Merge branch 'test_hhc' of https://github.com/timmyfaraday/HarmonicPowerModels.jl into test_hhc

timmyfaraday committed:
added all fairness principles

- reporting and start value needs to be added

timmyfaraday committed:
Merge branch 'test_hhc' of https://github.com/timmyfaraday/HarmonicPowerModels.jl into test_hhc

timmyfaraday committed:
report and start for fairness principles

timmyfaraday committed:
small changes

timmyfaraday committed:
final commit before merging and registering

timmyfaraday commented:
DO NOT DELETE THIS BRANCH
